proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+### Projecting Swift into .NET
+
+The majority of the work for Swift/.NET interop is determining how a type that exists in Swift should exist in .NET and what shape it should have. This section is a work in progress and will discuss how each feature in Swift will be projected into .NET, particularly in cases where there is not a corresponding .NET or C# language feature. Each feature should have a subheading describing how the projection will look and how any mechanisms to make it work will be designed.
+
+All designs in this section should be designed such that they are trimming and AOT-compatible by construction. We should work to ensure that no feature requires whole-program analysis (such as custom steps in the IL Linker) to be trim or AOT compatible.
+
+#### Swift to .NET Language Feature Projections
+
+TBD
+
+#### Projection Tooling Components
+
+The projection tooling should be split into these components:
+
+##### Importing Swift into .NET
+
+1. A tool that takes in a `.swiftinterface` file or Swift sources and produces C# code.
+2. A library that provides the basic support for Swift interop that the generated code builds on.
+3. User tooling to easily generate Swift projections for a given set of `.framework`s.
+4. (optional) A .NET shared framework package when targeting macOS, Mac Catalyst, iOS, or tvOS platforms that exposes the platform APIs or NuGet packages for each `.framework` that is exposed from Swift to .NET.
+    - This would be required to provide a single source of truth for Swift types so they can be exposed across an assembly boundary.
+
+#####  Exporting .NET to Swift
+
+There are two components to exporting .NET to Swift: Implementing existing Swift types in .NET and passing instances of those types to Swift, and exposing novel types from .NET code to Swift code to be created from Swift. Exposing novel types from .NET code to Swift code is considered out of scope at this time.
+
+For implementing existing Swift types in .NET, we will require the following tooling:
+
+1. A Roslyn source generator to generate any supporting code needed to produce any required metadata to pass instances of Swift-type-implementing .NET types to Swift.
```

Comment:
This is one way of doing this, but not the only way. One could also use the Swift details as the source of truth, emit source from a small tool and then include those during the compilation step. There is no requirement for a Roslyn source generator here. In fact, doing everything in a small tool ensures that both sides are the same.

I agree this could be done in a Roslyn source generator, but I do not think it is a requirement. Determining the UX needs for the Swift interop will help focus this decision point.

Reply:
I believe there are both pros and cons; If it is implemented as a source generator, it is closer to the runtime and is platform-agnostic. On the other hand, if it is integrated into projection tools, it is probably faster but is more platform-specific.

Reply:
> If it is implemented as a source generator, it is closer to the runtime and is platform-agnostic.

Any small stand alone tool can be pure C# and ship just like a source generator. This is merely about packaging logic, nothing more. There is no requirement that the tooling for this live in the Roslyn process.